Geometry_Engine: Add Centroid method for outline curves(s) with opening curve(s) in it.

LMarkowski opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #1705 

<!-- Add short description of what has been fixed -->
<s>Work in progress. Draft version of dealing with redundant code while querying centroids of planar shapes with openings.
Details in the parent issue.</s>
Adding a `Geometry_Engine` method querying a combine centroid of any set of planar, coplanar, non-self-intersecting and closed curves with optional openings. Method itself is rather not meant to be widely used due to lack of checks. Needed to be public for being used in other methods for objects like `IElement2D` or `IProfile`

### Test files
<!-- Link to test files to validate the proposed changes -->
[Old test file](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?viewid=a5a93cba%2Dfcca%2D46cc%2Da31a%2D80b8d5cbfd7b&id=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F01%5FTest%20Scripts%2FBHoM%5FEngine%2FGeometry%5FEngine%2FGeometry%5FEngine%2DIssue%231705%20Centroid%20from%20outline%20curves%28s%29%20with%20opening%20curve%28s%29%20in%20it)
[Geometry_Engine](https://burohappold.sharepoint.com/:f:/s/BHoM/Esd2eZj9Hc1Mkh5tar8m-NsBlIxJtyJ40zXatKwF4D3uOQ?e=2pvfb9) - general test for Centroid
[Spatial_Engine](https://burohappold.sharepoint.com/:f:/s/BHoM/EjAN0bkxublBirJzGJFTg8YBZXjfLC81kOPcpnTBXe_8vA?e=T7Xmj1) - general test for Centroid


### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->
- Added new `Centroid(this IEnumerable<ICurve> outlines, IEnumerable<ICurve> openings)` method in the `Geometry_Engine`.
- Added new `Centroid(PlanarSurface)` method in the `Geometry_Engine`.
- Updated `Centroid(IElement2D)` method in the `Spatial_Engine` to use new method from the `Geometry_Engine` and take tolerance as an input.
- Added descriptions for `Centroid()` methods.

### Additional comments
<!-- As required -->
- `[PreviousVersion]` attribute has been used to update `Centroid(IElement2D)` inputs so rebuilding Versioning_Toolkit is needed for scripts using outdated methods to work.
- Centroid method for `IProfie` interface will be implemented separately as soon as #1986 will get resolved.

IsakNaslundBh commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -26,16 +26,44 @@
+        /***************************************************/
+
+
+        public static Point Centroid(this IProfile profile, double tolerance = Tolerance.Distance)
+        {
+            return Centroid(profile.Edges, new List<ICurve>(), tolerance);
+        }
+
```
Do not think we should place this here. Even if profiles right now live in Geometry_oM they are planned to be migrated to Spation_oM, (see https://github.com/BHoM/BHoM/issues/756).

(Also, the profile edges would need to be checked for inner/outer as well, but that is another point)

pawelbaran replied:
Plus this method assumes that all the profile does not have holes - what about RHS for example?

IsakNaslundBh commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -26,16 +26,44 @@
+
+        /***************************************************/
+
+        //TODO: Move to structure_engine probably
+
+        //public static Point Centroid(this Panel panel, double tolerance = Tolerance.Distance)
+        //{
+        //    return Centroid(panel.ExternalEdges.Select(e => e.Curve).ToList(), panel.Openings.Select(o => o.Edges.Select(e => e.Curve)).SelectMany(o => o).ToList(), tolerance);
+        //}
+
```
Agree, this should not be in here. Should also probably make sure this works on IPanel/IElement2d

IsakNaslundBh commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -204,48 +231,76 @@ public static Point Centroid(this PolyCurve curve, double tolerance = Tolerance.
         /**** Private methods                           ****/
         /***************************************************/
 
+        private static Point Centroid(this IEnumerable<ICurve> outlines, IEnumerable<ICurve> openings = null, double tolerance = Tolerance.Distance)
```
Given the comments above, feel we almost need to make this public, to be able to reach it from other projects.

Wonder if some system with including checks in here, and having a boolean input stating whether the checks should be preformed or not could be a solution... Not sure on this last part, but it will probably need to be made public.

pawelbaran replied:
Agreed with @IsakNaslundBh, as in [my comment](https://github.com/BHoM/BHoM_Engine/issues/1705#issuecomment-671462380) I would even think of adding a description that this method requires closed and coplanar curves instead of any actual checks, moving the responsibility to the one that calls that method.

LMarkowski commented:
WIP. Changes made:
- Added descriptions to methods
- Made `Centroid(outlines, openings)` public
- Made already existing `IElement2D` method using new core method (logic behind both was roughly identical)
- Removed `Panel` method as it is covered by `IElement2D`
- Moved `IProfile` method to the `Spatial_Engine`

ToDo:
- Implement methods querying outer and inner edges for `IProfile` or remove `IProfile` method
- Create proper test files

@IsakNaslundBh @pawelbaran what are your opinions on current solution?

pawelbaran commented:
I like it, would say let's test and debug what has been done up to date and merge before the end of this sprint, with an option to push `Profile` query to the next Milestone (with possibly a different approach based on integration developed to calculate the section properties). What do you think @IsakNaslundBh?

LMarkowski commented:
@pawelbaran @IsakNaslundBh Added test files, removed method for IProfile and updated PR description.
Ready for final reviews! 🙌

pawelbaran reviewed (changes requested):
In general LGTM, but have spotted a few inconsistencies in the descriptions - please see the code comments.

Btw, good to see my `RandomClosedNonSelfIntersectingPlanarPolyCurves` still in use 🤣

LMarkowski commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -26,16 +26,72 @@
+        [Description("Queries the centre of area for a PlanarSurface.")]
+        [Input("planar surface", "The PlanarSurface to get the centre of area of.")]
+        [Input("tolerance", "Distance tolerance, default set to BH.oM.Geometry.Tolerance.Distance")]
+        [Output("centroid", "The Point at the centre of given PlanarSurface.")]
```
It is mentioned in IElement2D description. I went with the concept that geometrical objects do not have mass and therefore centre of mass. In opposite to analytical objects.

LMarkowski commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -26,16 +26,72 @@
+        //TODO: Remove when PlanarSurface will implement IElement2D
+        [Description("Queries the centre of area for a PlanarSurface.")]
+        [Input("planar surface", "The PlanarSurface to get the centre of area of.")]
+        [Input("tolerance", "Distance tolerance, default set to BH.oM.Geometry.Tolerance.Distance")]
```
I'm good with both solutions. Waiting for more opinions 😃

IsakNaslundBh replied:
Might as well update to pawels suggestion, as it gives a bit more context.

FraserGreenroyd commented:
> > good to see my `RandomClosedNonSelfIntersectingPlanarPolyCurves` still in use
> 
> @pawelbaran, If it ain't broken, don't fix it. 😉

So when are we fixing @pawelbaran 😆

pawelbaran commented:
> > > good to see my `RandomClosedNonSelfIntersectingPlanarPolyCurves` still in use
> > 
> > 
> > @pawelbaran, If it ain't broken, don't fix it. 😉
> 
> So when are we fixing @pawelbaran 😆

But it is not broken 🙊

pawelbaran commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -234,6 +302,10 @@ public static Point Centroid(this Line line, double tolerance = Tolerance.Distan
         /***************************************************/
 
+        [Description("Interface method that queries the centre of area for any ICurve.")]
+        [Input("iCurve", "The ICurve to get the centre of area of.")]
```
```suggestion
        [Input("curve", "The ICurve to get the centre of area of.")]
```

pawelbaran commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -95,11 +149,14 @@ public static Point Centroid(this Polyline curve, double tolerance = Tolerance.D
+        [Description("Queries the centre of area enclosed by a closed, planar, non-self-intersecting PolyCurve.")]
+        [Input("curve", "The PolyCurve to get the centre of area of.")]
+        [Input("tolerance", "Distance tolerance, default set to BH.oM.Geometry.Tolerance.Distance")]
+        [Output("centroid", "The Point at the centre of area enclosed by given Polyline.")]
```
```suggestion
        [Output("centroid", "The Point at the centre of area enclosed by given PolyCurve.")]
```

LMarkowski replied:
🤦

pawelbaran reviewed (changes requested):
Two minor description inconsistencies found, besides that I am happy as soon as @IsakNaslundBh is happy 👍

pawelbaran reviewed (dismissed):
LGTM - hitting approve but not entirely sure about `tolerance` description - @IsakNaslundBh could you please have a quick look before merging?

IsakNaslundBh commented on Geometry_Engine/Query/Centroid.cs:
```diff
@@ -26,16 +26,72 @@
+                area += outlineArea;
+            }
+
+            foreach (ICurve opening in openings.BooleanUnion(tolerance))
```
need to null check the list of openings before this is being hit. Especially important as the method allows for default nulls as input. something like `openings  = openings ?? new List<ICurve>();` before this is run could work.

IsakNaslundBh reviewed (changes requested):
null check needed on opening curves

LMarkowski commented:
/azp run BHoM_Engine.CheckInstaller

IsakNaslundBh reviewed (dismissed):
Now looks good!

IsakNaslundBh commented on Spatial_Engine/Query/Centroid.cs:
```diff
@@ -54,11 +54,11 @@ public static Point Centroid(this IElement1D element1D)
         public static Point Centroid(this IElement2D element2D, double tolerance = Tolerance.Distance)
         {
-            return Geometry.Query.Centroid(new List<ICurve> { element2D.OutlineCurve() }, element2D.InternalOutlineCurves(), tolerance);
+            return Geometry.Query.Centroid(new List<ICurve> { element2D.OutlineCurve() });//, element2D.InternalOutlineCurves(), tolerance);
```
this commented out code accidentally got commited from testing I presume!

LMarkowski replied:
yeaaa 😅

IsakNaslundBh commented:
/azp run BHoM_Engine.CheckInstaller

IsakNaslundBh reviewed (approved):
now, LGTM
